providers/apis/defi/uniswapv3/fetcher.go

```diff
@@ -0,0 +1,269 @@
+	}
+
+	if metrics == nil {
+		return nil, fmt.Errorf("metrics cannot be nil")
+	}
+
+	if api.Name != Name {
+		return nil, fmt.Errorf("expected api config name %s, got %s", Name, api.Name)
+	}
+
+	if !api.Enabled {
+		return nil, fmt.Errorf("api config for %s is not enabled", Name)
+	}
+
+	if err := api.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("invalid api config: %w", err)
+	}
+
+	abi, err := uniswappool.UniswapMetaData.GetAbi()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get uniswap abi: %w", err)
+	}
+
+	payload, err := abi.Pack(ContractMethod)
+	if err != nil {
+		return nil, fmt.Errorf("failed to pack slot0: %w", err)
+	}
+
+	return &PriceFetcher{
+		logger:    logger,
+		metrics:   metrics,
+		api:       api,
+		client:    client,
+		abi:       abi,
+		payload:   payload,
+		poolCache: make(map[mmtypes.Ticker]PoolConfig),
+	}, nil
+}
+
+// Fetch returns the price of a given set of tickers. This fetch utilizes the batch call to lower
+// overhead of making individual RPC calls for each ticker. The fetcher will query the Uniswap V3
+// pool contract for the price of the pool. The price is derived from the slot 0 data of the pool
+// contract, specifically the sqrtPriceX96 value.
+func (u *PriceFetcher) Fetch(
+	ctx context.Context,
+	tickers []mmtypes.Ticker,
+) types.PriceResponse {
+	start := time.Now()
+	defer func() {
+		u.metrics.ObserveProviderResponseLatency(Name, time.Since(start))
+	}()
+
+	var (
+		resolved   = make(types.ResolvedPrices)
+		unResolved = make(types.UnResolvedPrices)
+	)
+
+	// Create a batch element for each ticker and pool.
+	batchElems := make([]rpc.BatchElem, len(tickers))
+	pools := make([]PoolConfig, len(tickers))
+	for i, ticker := range tickers {
+		pool, err := u.GetPool(ticker)
+		if err != nil {
+			u.logger.Error(
+				"failed to get pool for ticker",
+				zap.String("ticker", ticker.String()),
+				zap.Error(err),
+			)
+
+			return types.NewPriceResponseWithErr(
+				tickers,
+				providertypes.NewErrorWithCode(
+					fmt.Errorf("failed to get pool: %w", err),
+					providertypes.ErrorFailedToDecode,
+				),
+			)
+		}
+
+		// Create a batch element for the ticker and pool.
+		var result string
+		batchElems[i] = rpc.BatchElem{
+			Method: "eth_call",
+			Args: []interface{}{
+				map[string]interface{}{
+					"to":   common.HexToAddress(pool.Address),
+					"data": hexutil.Bytes(u.payload), // slot0 call to the pool contract.
+				},
+				"latest", // latest signifies the latest block.
+			},
+			Result: &result,
+		}
+		pools[i] = pool
+	}
+
+	// Batch call to the EVM.
+	if err := u.client.BatchCallContext(ctx, batchElems); err != nil {
+		u.logger.Error(
+			"failed to batch call to ethereum network for all tickers",
+			zap.Error(err),
+		)
+
+		return types.NewPriceResponseWithErr(
+			tickers,
+			providertypes.NewErrorWithCode(err, providertypes.ErrorAPIGeneral),
+		)
+	}
+
+	// Parse the result from the batch call for each ticker.
+	for i, ticker := range tickers {
```

Comment:
Is there a way to guarantee the order of responses from batch-json RPC requests?

Reply:
the client should return things in the same order, but we can probably resort just for sanity before using the responses.